inc/ad-servers/class-ad-layers-dfp.php

```diff
@@ -0,0 +1,1176 @@
+					( ! empty( $this->mapping_by_unit[ $ad_unit ] ) && ! in_array( $ad_unit, $this->oop_units, true ) ) ? '.defineSizeMapping(dfpBuiltMappings[' . wp_json_encode( $ad_unit ) . '])' : '',
+					( ! empty( $this->targeting_by_unit[ $ad_unit ] ) ) ? $this->targeting_by_unit[ $ad_unit ] : '' // This is escaped above as it is built.
+				);
+				// phpcs:enable
+			}
+		}
+
+		/**
+		 * Get the ad units for the given ad layer id.
+		 *
+		 * @param  int $ad_layer_id ad-layer post ID.
+		 * phpcs:disable VariableAnalysis.CodeAnalysis.VariableAnalysis.UndefinedVariable
+		 * @return array
+		 */
+		public function get_ad_units_for_layer( $ad_layer_id ) {
+			$global_ad_units = $this->get_setting( 'ad_units' );
+			foreach ( $global_ad_units as $global_ad_unit ) {
+				// TODO: $global_path_overrides doesn't exist.
+				if ( ! empty( $global_ad_unit['code'] ) && ! empty( $global_ad_unit['path_override'] ) ) {
+					$global_path_overrides[ $global_ad_unit['code'] ] = $global_ad_unit['path_override'];
+				}
+			}
+
+			$this->ad_units = [];
+			$temp_ad_units  = get_post_meta( $ad_layer_id, 'ad_layer_ad_units', true );
+			if ( ! empty( $temp_ad_units ) ) {
+				foreach ( $temp_ad_units as $ad_unit ) {
+					if ( ! empty( $ad_unit['ad_unit'] ) ) {
+						if ( ! isset( $ad_unit['custom_targeting'] ) ) {
+							$ad_unit['custom_targeting'] = [];
+						}
+						$this->ad_units[ $ad_unit['ad_unit'] ] = $ad_unit['custom_targeting'];
+						if ( ! empty( $ad_unit['do_not_render'] ) ) {
+							$this->do_not_render_ads[ $ad_unit['ad_unit'] ] = true;
+						}
+						if ( ! empty( $ad_unit['path_override'] ) ) {
+							$this->ad_unit_paths[ $ad_unit['ad_unit'] ] = $ad_unit['path_override'];
+						} elseif ( ! empty( $global_path_overrides[ $ad_unit['ad_unit'] ] ) ) {
+							$this->ad_unit_paths[ $ad_unit['ad_unit'] ] = $global_path_overrides[ $ad_unit['ad_unit'] ];
+						}
+					}
+				}
+			}
+
+			return $this->ad_units;
+			/* phpcs:enable */
+		}
+
+		/**
+		 * Creates the DFP targeting Javascript.
+		 *
+		 * @access private
+		 * @param array $ad_layer ad layer config array.
+		 */
+		private function targeting_js( $ad_layer ) {
+			// Handle any page level custom targeting specified for this ad layer.
+			$custom_targeting = get_post_meta( $ad_layer['post_id'], 'ad_layer_custom_targeting', true );
+			$custom_targeting = apply_filters( 'ad_layers_dfp_page_level_targeting', $custom_targeting );
+
+			if ( empty( $custom_targeting ) ) {
+				return;
+			}
+
+			// Add the JS.
+			if ( ! empty( $custom_targeting ) ) {
+				echo 'googletag.pubads()' . apply_filters( 'ad_layers_dfp_page_level_targeting_output_html', $this->get_targeting_js_from_array( $custom_targeting ) ) . ";\n"; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+			}
+		}
+
+		/**
+		 * Get the main details for all ads for use by the JS API.
+		 *
+		 * @return array
+		 */
+		protected function get_ad_details() {
```

Comment:
Lots of methods could be type cased since you are using PHP 7.4. 

https://github.com/alleyinteractive/ad-layers/blob/75bd3d57a42f131ae882de97863e94fc3717141e/.phpcs.xml could also be updated to reflect that PHP version.

```suggestion
		protected function get_ad_details(): array {
```

Reply:
This is a good point. We were supporting 5.8 because this is an open-source plugin, I assume. I might figure out what options we have there.